<a href="https://colab.research.google.com/github/matalhim/mephi/blob/master/Binding_energy_of_the_nucleus.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Энергия связи ядра (формула Вайцзеккера)
$ W(A,Z) = \alpha A - \beta A^{2/3} - \gamma \frac{Z(Z-1)}{A^{1/3}} - \delta \frac{(A-2Z)^2}{A} + \zeta A^{-3/4}$

$^3\text{H},\ ^5\text{He},\ ^7\text{Li},\ ^9\text{B},\ ^{11}\text{B},\ ^{11}\text{C},\ ^{13}\text{C},\ ^{13}\text{N},\ ^{15}\text{N},\ ^{17}\text{O},\ ^{19}\text{Ne},\ ^{23}\text{Na},\ ^{23}\text{Mg},\ ^{25}\text{Mg},\ ^{27}\text{Al},\ ^{27}\text{Si},\ ^{29}\text{Si},\ ^{31}\text{Si},\ ^{33}\text{S},\ ^{35}\text{Cl},\ ^{37}\text{Ar},\ ^{39}\text{K},\ ^{45}\text{Sc},\ ^{55}\text{Mn},\ ^{59}\text{Fe},\ ^{65}\text{Ni},\ ^{63}\text{Cu},\ ^{65}\text{Cu},\ ^{65}\text{Zn},\ ^{103}\text{Rh},\ ^{107}\text{Ag},\ ^{143}\text{Ba},\ ^{141}\text{Pr},\ ^{197}\text{Au},\ ^{207}\text{Pb},\ ^{233}\text{Th},\ ^{235}\text{Pa},\ ^{239}\text{U},\ ^{239}\text{Pu}$


\begin{array}{|c|c|c|c|}
\hline
\text{№} & \text{Элемент} & A & Z \\
\hline
1 & ^3\text{H} & 3 & 1 \\
2 & ^5\text{He} & 5 & 2 \\
3 & ^7\text{Li} & 7 & 3 \\
4 & ^9\text{B} & 9 & 4 \\
5 & ^{11}\text{B} & 11 & 5 \\
6 & ^{11}\text{C} & 11 & 6 \\
7 & ^{13}\text{C} & 13 & 6 \\
8 & ^{13}\text{N} & 13 & 7 \\
9 & ^{15}\text{N} & 15 & 7 \\
10 & ^{17}\text{O} & 17 & 8 \\
11 & ^{19}\text{Ne} & 19 & 10 \\
12 & ^{23}\text{Na} & 23 & 11 \\
13 & ^{23}\text{Mg} & 23 & 12 \\
14 & ^{25}\text{Mg} & 25 & 12 \\
15 & ^{27}\text{Al} & 27 & 13 \\
16 & ^{27}\text{Si} & 27 & 14 \\
17 & ^{29}\text{Si} & 29 & 14 \\
18 & ^{31}\text{Si} & 31 & 14 \\
19 & ^{33}\text{S} & 33 & 16 \\
20 & ^{35}\text{Cl} & 35 & 17 \\
21 & ^{37}\text{Ar} & 37 & 18 \\
22 & ^{39}\text{K} & 39 & 19 \\
23 & ^{45}\text{Sc} & 45 & 21 \\
24 & ^{55}\text{Mn} & 55 & 25 \\
25 & ^{59}\text{Fe} & 59 & 26 \\
26 & ^{65}\text{Ni} & 65 & 28 \\
27 & ^{63}\text{Cu} & 63 & 29 \\
28 & ^{65}\text{Cu} & 65 & 29 \\
29 & ^{65}\text{Zn} & 65 & 30 \\
30 & ^{103}\text{Rh} & 103 & 45 \\
31 & ^{107}\text{Ag} & 107 & 47 \\
32 & ^{143}\text{Ba} & 143 & 56 \\
33 & ^{141}\text{Pr} & 141 & 59 \\
34 & ^{197}\text{Au} & 197 & 79 \\
35 & ^{207}\text{Pb} & 207 & 82 \\
36 & ^{233}\text{Th} & 233 & 90 \\
37 & ^{235}\text{Pa} & 235 & 91 \\
38 & ^{239}\text{U} & 239 & 92 \\
39 & ^{239}\text{Pu} & 239 & 94 \\
\hline
\end{array}

In [55]:
a_values = [3, 5, 7, 9, 11, 11, 13, 13, 15, 17, 19, 23, 23, 25, 27, 27, 29, 31, 33, 35, 37, 39, 45, 55, 59, 65, 63, 65, 65, 103, 107, 143, 141, 197, 207, 233, 235, 239, 239]
z_values = [1, 2, 3, 4, 5, 6, 6, 7, 7, 8, 10, 11, 12, 12, 13, 14, 14, 14, 16, 17, 18, 19, 21, 25, 26, 28, 29, 29, 30, 45, 47, 56, 59, 79, 82, 90, 91, 92, 94]
m_values= [3.016049, 5.012057, 7.016004, 9.013330, 11.009305, 11.011433, 13.003355, 13.005739, 15.000109, 16.999132, 19.001880, 22.989769, 22.994124, 24.985837, 26.981538, 26.986705,
           28.976495, 30.975363, 32.971459, 34.968853, 36.966776, 38.963706, 44.955908, 54.938043, 58.934874, 64.930085, 62.929597, 64.927790, 64.929241,
           102.905494, 106.905092, 142.920625, 140.907658, 196.966570, 206.975897, 233.041580, 235.045399, 239.054292, 239.052162]

$E_{св}(A,Z)= Zm_{p}+(A-Z)m_{n}-M(A,Z)$


Энергия связи ядра где $Z$ - число протонов, $(A - Z)$ - число нейтронов, $m_p$ - масса протона, $m_n$ - масса нейтрона, $М(A,Z)$ - масса ядра с массовым числом $A$ и зарядом $Z$.



In [56]:
m_p = 938.272
m_n = 931.565
a_e_m = 931.5

In [64]:
def calculate_true_E(a, z, m):
    return z * m_p + (a - z) * m_n - m * a_e_m

e_true_values = [calculate_true_E(a_values[i], z_values[i], m_values[i]) for i in range(len(a_values))]
for i in range(len(e_true_values)):
    print('{}: {}'.format(i+1, e_true_values[i]))

1: -8.04764350000005
2: 2.5079045000002225
3: 5.668274000001475
4: 14.99610500000017
5: 25.582392500000424
6: 30.307160500002283
7: 37.9618174999996
8: 42.448121500001434
9: 47.82246650000161
10: 55.56954200000291
11: 66.553780000002
12: 84.80217650000122
13: 87.45249400000102
14: 95.3018345000055
15: 106.14335299999948
16: 108.03729250000106
17: 117.67790750000495
18: 118.8623654999974
19: 136.04294149999987
20: 145.30743049999728
21: 154.07915599999978
22: 163.77586100000917
23: 184.84369799999695
24: 228.96294550000312
25: 238.88186900000437
26: 257.14682249999896
27: 264.1783945000061
28: 265.991614999999
29: 271.34700849999354
30: 396.5423390000069
31: 410.5908020000061
32: 458.8248124999809
33: 490.89457299999776
34: 573.7980450000032
35: 585.8809445000079
36: 580.0432300000102
37: 583.3228315000306
38: 582.0060020000092
39: 597.4040970000206


Посчитаны истинные значения энергий для элементов из таблицы

Далее, используя метод наименьших квадратов, ищется минимум функционала
следующего вида:
$F = (\sum_{i} E_{true\ i} - \alpha \sum_{i}A_{i} + \beta \sum_{i}A^{2/3}_{i} + \gamma \sum_{i}\frac{Z_{i}(Z_{i}-1)}{A_{i}^{1/3}} + \delta \sum_{i}\frac{(A_{i}-2Z_{i})^2}{A_{i}} - \zeta \sum_{i}A{i}^{-3/4})^2$ \\
Условием поиска минимума функционала является приравнивание к нулю его
частных производных по неизвестным параметрам $\alpha,\ \beta,\ \gamma, \ \delta$:


\begin{cases}
        \frac{\partial F}{\partial \alpha} = 0,\\
        \frac{\partial F}{\partial \beta} = 0,\\
        \frac{\partial F}{\partial \gamma} = 0,\\
        \frac{\partial F}{\partial \delta} = 0,\\
    \end{cases}\
    \begin{cases}
        \alpha \sum_{i}A_{i} \cdot A_{i} - \beta \sum_{i}A_{i}^{2/3}\cdot A_{i} - \gamma \sum_{i}\frac{Z_{i}(Z_{i}-1)}{A_{i}^{1/3}} \cdot A_{i} - \delta \sum_{i}\frac{(A_{i}-2Z_{i})^2}{A_{i}} \cdot A_{i}  \\= \sum_{i}E_{true\ i} \cdot A_{i} - \zeta \sum_{i}A_{i}^{-3/4} \cdot A_{i}\\
        \\
        \alpha \sum_{i}A_{i} \cdot A_{i}^{2/3} - \beta \sum_{i}A_{i}^{2/3}\cdot A_{i}^{2/3} - \gamma \sum_{i}\frac{Z_{i}(Z_{i}-1)}{A_{i}^{1/3}} \cdot A_{i}^{2/3} - \delta \sum_{i}\frac{(A_{i}-2Z_{i})^2}{A_{i}} \cdot A_{i}^{2/3}   \\= \sum_{i}E_{true\ i} \cdot A_{i}^{2/3} - \zeta \sum_{i}A_{i}^{-3/4} \cdot A_{i}^{2/3}\\
        \\
        \alpha \sum_{i}A_{i} \cdot \frac{Z_{i}(Z_{i}-1)}{A_{i}^{1/3}} - \beta \sum_{i}A_{i}^{2/3}\cdot \frac{Z_{i}(Z_{i}-1)}{A_{i}^{1/3}} - \gamma \sum_{i}\frac{Z_{i}(Z_{i}-1)}{A_{i}^{1/3}} \cdot \frac{Z_{i}(Z_{i}-1)}{A_{i}^{1/3}} - \delta \sum_{i}\frac{(A_{i}-2Z_{i})^2}{A_{i}} \cdot \frac{Z_{i}(Z_{i}-1)}{A_{i}^{1/3}} \\= \sum_{i}E_{true\ i} \cdot \frac{Z_{i}(Z_{i}-1)}{A_{i}^{1/3}} - \zeta \sum_{i}A_{i}^{-3/4} \cdot \frac{Z_{i}(Z_{i}-1)}{A_{i}^{1/3}} \\
        \\
        \alpha \sum_{i}A_{i} \cdot A_{i}^{-3/4} - \beta \sum_{i}A_{i}^{2/3}\cdot A_{i}^{-3/4} - \gamma \sum_{i}\frac{Z_{i}(Z_{i}-1)}{A_{i}^{1/3}} \cdot A_{i}^{-3/4} - \delta \sum_{i}\frac{(A_{i}-2Z_{i})^2}{A_{i}} \cdot A_{i}^{-3/4}  \\= \sum_{i}E_{true\ i} \cdot A_{i}^{-3/4} - \zeta \sum_{i}A_{i}^{-3/4} \cdot A_{i}^{-3/4}\\
    \end{cases}

Система уравнений может быть переписана в матричном виде:


$\vec{\Omega} = M\times\vec{\Theta}$

где:
$\vec{\Theta} = \begin{pmatrix}
\alpha \\
\beta\\
\gamma \\
\delta \\
\end{pmatrix},
\vec{\Omega} = \begin{pmatrix}
\sum_{i}(E_{true\ i} - \zeta A_{i}^{-3/4}) \cdot x_{i} \\
\sum_{i}(E_{true\ i} - \zeta A_{i}^{-3/4}) \cdot y_{i} \\
\sum_{i}(E_{true\ i} - \zeta A_{i}^{-3/4}) \cdot z_{i} \\
\sum_{i}(E_{true\ i} - \zeta A_{i}^{-3/4}) \cdot \xi_{i} \\
\end{pmatrix},
$

$
M = \begin{pmatrix}
\sum_{i} x_{i}^{2} & -\sum_{i} x_{i}\cdot y_{i} & -\sum_{i}x_{i}\cdot z_{i} & -\sum_{i}x_{i}\cdot \xi_{i} \\
\sum_{i}x_{i}\cdot y_{i} & -\sum_{i} y_{i}^{2} & -\sum_{i}y_{i}\cdot z_{i} & -\sum_{i}y_{i}\cdot \xi_{i} \\
\sum_{i}x_{i}\cdot z_{i} & -\sum_{i}y_{i}\cdot z_{i} & -\sum_{i} z_{i}^{2} & -\sum_{i}z_{i}\cdot \xi_{i} \\
\sum_{i}x_{i}\cdot \xi_{i} & -\sum_{i}y_{i}\cdot \xi_{i} & -\sum_{i}z_{i}\cdot \xi_{i} &-\sum_{i}\xi_{i}^{2} \\
\end{pmatrix}
$

где:
$
x_{i} = A_{i}, \ y_{i} = A^{2/3}_{i}, \ z_{i} = \frac{Z_{i}(Z_{i}-1)}{A_{i}^{1/3}}, \ \xi_{i} = \frac{(A_{i}-2Z_{i})^2}{A_{i}}
$

In [58]:
def calculate_x_i(a):
    return a

def calculate_y_i(a):
    return a ** (2/3)

def calculate_z_i(a, z):
    return z * (z-1) /(a ** (1/3))

def calculate_xi_i(a, z):
    return ((a-2*z) ** 2) / a

In [59]:
omega = [0 for _ in range(4)]

def calculate_zeta(a, z):
    if a %2 == 0 and a %2 == 0:
        return 34
    elif a %2 != 0 and a %2 != 0:
        return -34
    else:
        return 0

for i in range(len(e_true_values)):
    x = calculate_x_i(a_values[i])
    y = calculate_y_i(a_values[i])
    z = calculate_z_i(a_values[i], z_values[i])
    xi = calculate_xi_i(a_values[i], z_values[i])
    zeta = calculate_zeta(a_values[i], z_values[i])

    omega[0] += (e_true_values[i] - zeta * a_values[i] ** (-3/4)) * x
    omega[1] += (e_true_values[i] - zeta * a_values[i] ** (-3/4)) * y
    omega[2] += (e_true_values[i] - zeta * a_values[i] ** (-3/4)) * z
    omega[3] += (e_true_values[i] - zeta * a_values[i] ** (-3/4)) * xi
print(omega)

[1160402.2886780808, 215516.02700982708, 5789012.17643005, 44955.7234221758]


In [60]:
import numpy as np

m = np.zeros((4, 4))

for i in range(len(e_true_values)):
    x = calculate_x_i(a_values[i])
    y = calculate_y_i(a_values[i])
    z = calculate_z_i(a_values[i], z_values[i])
    xi = calculate_xi_i(a_values[i], z_values[i])

    m[0][0] += x ** 2
    m[0][1] += - x * y
    m[0][2] += - x * z
    m[0][3] += - x * xi


    m[1][1] += - y ** 2
    m[1][2] += - y * z
    m[1][3] += - y * xi

    m[2][2] += - z ** 2
    m[2][3] += - z * xi

    m[3][3] += - xi ** 2
m[1][0] = - m[0][1]
m[2][0] = - m[0][2]
m[3][0] = - m[0][3]
m[2][1] =  m[1][2]
m[3][1] =  m[1][3]
m[3][2] =  m[2][3]
print(m)



[[ 4.04743000e+05 -7.28092776e+04 -2.08494811e+06 -1.67350000e+04]
 [ 7.28092776e+04 -1.36419406e+04 -3.61849616e+05 -2.82712216e+03]
 [ 2.08494811e+06 -3.61849616e+05 -1.10914152e+07 -9.08382785e+04]
 [ 1.67350000e+04 -2.82712216e+03 -9.08382785e+04 -7.74359913e+02]]


Решением данного матричного уравнения является:


$\vec{\Theta} = M^{-1} \times \vec{\Omega}$

In [66]:
inverse_m = np.linalg.inv(m)
omega_vector = np.array(omega)
theta = np.dot(inverse_m, omega_vector)
print(theta)

[ 9.56459651 12.70106805  0.61093747 30.61085645]


Получилось:

$\alpha = 9.56459651, \ \alpha_{true} =15.6$

$\beta = 12.70106805, \ \beta_{true} = 17.2$

$\gamma = 0.61093747, \ \gamma_{true} = 0.72$

$\delta = 30.61085645, \ \delta_{true} = 23.6$

Расходимость с истинными значениями скорее всего свяазана с ошибкой вычисления корней уравнения.